### PR TITLE
Stacked Bar: Summed Labels

### PR DESCRIPTION
**Pull request by CalvinFernandez**
image
image
image

### CODE REVIEW

File: src/utils/stackingUtils.ts [hunk 1]
```diff
@@ -46,6 +46,33 @@ namespace Plottable.Utils.Stacking {
...
   }
 
+  export function stackedExtents(stackingResult: StackingResult) {
+    const maximumExtents: Utils.Map<string, number> = new Utils.Map<string, number>();
+    const minimumExtents: Utils.Map<string, number> = new Utils.Map<string, number>();
+
+    stackingResult.forEach((stack) => {
+      stack.forEach((datum: StackedDatum, key: string) => {
+        // correctly handle negative bar stacks
+        const maximalValue = Math.max([datum.offset + datum.value, datum.offset], datum.offset);
```
Comment (hellochar) [inline src/utils/stackingUtils.ts:71]: How does this work? It looks like we're passing an array argument to Math.max, which shouldn't work

File: src/utils/stackingUtils.ts [hunk 2]
```diff
@@ -46,6 +46,33 @@ namespace Plottable.Utils.Stacking {
     return datasetToKeyToStackedDatum;
   }
 
+  export function stackedExtents(stackingResult: StackingResult) {
+    const maximumExtents: Utils.Map<string, number> = new Utils.Map<string, number>();
+    const minimumExtents: Utils.Map<string, number> = new Utils.Map<string, number>();
+
+    stackingResult.forEach((stack) => {
+      stack.forEach((datum: StackedDatum, key: string) => {
```
Comment (hellochar) [inline src/utils/stackingUtils.ts]: nit: keys could be numbers or Dates, (e.g. Numeric x scale on histograms)

File: src/plots/stackedBarPlot.ts [hunk 3]
```diff
@@ -54,6 +60,92 @@ namespace Plottable.Plots {
...
+      this._writer = new SVGTypewriter.Writers.Writer(this._measurer);
+    }
+
+    protected _drawLabels() {
+      super._drawLabels();
+
+      // remove all current labels before redrawing
+      this._labelArea.selectAll("g").remove();
+
+      const baselineValue = +this.baselineValue();
```
Comment (hellochar) [inline src/plots/stackedBarPlot.ts:75]: Why do we convert to number here

**Review by hellochar — APPROVED** [review]
* can/should we DRY this with the barPlot label drawing code at all?

* would it be difficult to draw labels (or truncate) if the area is too small? I've seen some other charts where the bar stack itself is small (~10px) but the label just bleeds a bit into the other stacks and it looks fine

File: src/utils/stackingUtils.ts (hunk 2 above)
Comment (CalvinFernandez) [reply src/utils/stackingUtils.ts]: oops good point

File: src/utils/stackingUtils.ts (hunk 1 above)
Comment (CalvinFernandez) [reply src/utils/stackingUtils.ts:71]: This is the `Math.max` from Utils, not the window's Math.max

File: src/plots/stackedBarPlot.ts (hunk 3 above)
Comment (CalvinFernandez) [reply src/plots/stackedBarPlot.ts:75]: the baseline value isn't _technically_ numeric, but I'm operating under the assumption that you need a numeric secondary scale in order to have a stacked bar that works correctly

**Comment by pkwi** [comment]
With font size 12pt - it might fit for every individual bar that is minimum 10px high (See Sprint bar). Otherwise, text might not be readable (See Winter bar).

screen shot 2017-01-19 at 2 33 18 pm

**Comment by CalvinFernandez** [comment]
> Should we DRY? 

Yea, but I'm too scared to touch that code here. 

> would it be difficult to draw labels (or truncate) if the area is too small?

Probably wouldn't be too hard but truncated numbers lose more meaning than truncated strings so its really a question of whether it makes sense or not @pkwi thoughts? 

> I've seen some other charts where the bar stack itself is small (~10px) but the label just bleeds a bit into the other stacks and it looks fine

Yea we can probably finagle the width to consume 100% of the space. @pkwi thoughts?  Down to do this in a follow-up commit.

**Comment by CalvinFernandez** [comment]
image

**Comment by pkwi** [comment]
I would avoid allowing labels to stick outside the stack. Otherwise, we might have a situation like this:

screen shot 2017-01-19 at 6 04 28 pm

Where labels are overlapping each other.

**Comment by hellochar** [comment]
Are overlapping labels that bad? In the example you gave I can at least still see all the numbers and the bars they represent, which does convey more information

**Review by hellochar — APPROVED** [review]

**Comment by ccolgrove** [comment]
@hellochar Re: overlapping labels. This is bad because it _looks_ broken, regardless of whether or not it actually is broken

**Pull request merged**